### PR TITLE
AV-213944 Status update concurrency fix

### PR DESCRIPTION
**Pull request by aaha97**
fix incorrect status update due to concurrency issue.

the issue was found when running the UT ingresstests::TestMultiHostIngressStatusCheck

concurrent threads were overwriting valid IP in status because updates were not atomic in nature and both referred to the same old ingress objects.

this is fixed in this PR by adding locks before status updates when same object is being updated by multiple threads and fetching the latest status before proceeding with the update

### CODE REVIEW

**Comment by Dhivyaaj** [comment]
No JIRA Ids found for the PR. Jira id is mandatory to update fix version in jira.                                            Please update respective Jira id in PR title or commit message                                            if the PR is intented for default branches of repo.                                            For mandatory exemptions, comment trigger phrase 'skip jira-id-check' in PR.                                            For manual trigger, comment trigger phrase 'run jira-id-check' in PR.

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

File: internal/status/ing_status.go [hunk 1]
```diff
@@ -149,6 +149,19 @@ func updateObject(mIngress *networkingv1.Ingress, updateOption UpdateOptions, re
 	var updatedIng *networkingv1.Ingress
 	var err error
 	if !sameStatus {
+		ingressNsName := mIngress.Namespace + "/" + mIngress.Name
+		//lock here to avoid concurrent updates to same status
+		lib.GetLockSet().Lock(ingressNsName)
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply internal/status/ing_status.go:154]: ` getIngresses` is being called twice at 155 and 173. Is it possible to make it one call?

File: internal/lib/utils.go [hunk 2]
```diff
@@ -546,3 +547,31 @@ func IsInfraSettingNSScoped(infraSetting, namespace string) bool {
 	}
 	return false
 }
+
+type LockSet struct {
+	locks map[string]*sync.Mutex
```
Comment (abhishekbsingh) [inline internal/lib/utils.go:552]: we can use sync.map here and do away with explicit lock/unlock logic.

Comment (aaha97) [reply internal/lib/utils.go:552]: all threads are required to perform atomic updates. sync.map will not work for this use case

Comment (abhishekbsingh) [reply internal/lib/utils.go:552]: ok, in that case let's use sync.map instead of regular map which would not be safe for concurrent write/read

Comment (aaha97) [reply internal/lib/utils.go]: replaced regular golang map with sync.Map, thanks

**Comment by aaha97** [comment]
build ako

File: internal/lib/utils.go [hunk 3]
```diff
@@ -546,3 +547,26 @@ func IsInfraSettingNSScoped(infraSetting, namespace string) bool {
...
+var lockSet LockSet
+
+func (s *LockSet) Lock(lockName string) {
+	lock, _ := s.locks.LoadOrStore(lockName, &sync.Mutex{})
+	lock.(*sync.Mutex).Lock()
+}
+
+func (s *LockSet) Unlock(lockName string) {
+	if lock, ok := s.locks.Load(lockName); !ok {
+		panic("unlocked an unlock mutex")
```
Comment (abhishekbsingh) [inline internal/lib/utils.go:564]: If the lockName(ingress) is not there is the map - shouldn't the unlock operation be assumed successful? 
Why panic in that case?

Comment (aaha97) [reply internal/lib/utils.go:564]: the map will only get populated if the lock for ingress has been called atleast once. this panic should ideally never occur unless there is a code bug. 

we are never deleting the keys in the map, so the keys remain until AKO reboots.

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

**Comment by ashishnayak96** [comment]
build ako

**Comment by ashishnayak96** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

**Review by akshayhavile — APPROVED** [review]

**Comment by aaha97** [comment]
build ako

**Review by abhishekbsingh — APPROVED** [review]

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

**Comment by akshayhavile** [comment]
@aaha97 Can you please paste local UT run for this?

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
build ako

**Comment by aaha97** [comment]
UT results for ingresstests

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/ingresstests  1102.513s

**Pull request merged**